### PR TITLE
feat(test): adding an extended test for miscompare during partial rebuild

### PR DESCRIPTION
**Pull request by dsavitskiy**

### CODE REVIEW

**Comment by dsavitskiy** [comment]
bors test

File: ci.nix [hunk 1]
```diff
@@ -13,6 +13,7 @@ let
   norust_moth =
     "You have requested environment without RUST, you should provide it!";
   norustc_msg = "no rustc, use rustup tool to install it";
+  nofio_msg = "FIO cannot be found";
```
Comment (tiagolobocastro) [inline ci.nix]: why is this needed? doesn't fio come from the shell?

File: io-engine/tests/nexus_rebuild_partial_loop.rs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        }
+    });
+    tokio::pin!(j1);
+
+    println!("[C]");
+    println!("[C] Joining jobs ...");
+    let (io_res, _) = tokio::join!(j0, j1);
+    println!("[C] Joining jobs done");
+    let io_res = io_res.unwrap();
+    io_res.unwrap();
```
Comment (tiagolobocastro) [inline io-engine/tests/nexus_rebuild_partial_loop.rs:262]: how long does this run for if there's no miscompare?

File: io-engine-tests/src/fio.rs [hunk 3]
```diff
@@ -265,6 +281,8 @@ impl Fio {
         self.push_err(&stderr);
         self.exit = exit;
 
+        // println!("{stdout}");
```
Comment (dsharma-dc) [inline io-engine-tests/src/fio.rs]: Remove this now?

File: io-engine-tests/src/nexus.rs [hunk 4]
```diff
@@ -504,6 +504,50 @@ impl NexusBuilder {
         }
     }
 
+    pub async fn wait_children_online2(
```
Comment (dsharma-dc) [inline io-engine-tests/src/nexus.rs]: Why not add the extra string printing logic into existing `wait_children_online` function itself?

File: io-engine/tests/nexus_rebuild_partial_loop.rs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+    let j0 = tokio::spawn({
+        let nvmf = nex.nvmf_location();
+        async move {
+            println!("[A]");
+            let (_cg, path) = nvmf.open().unwrap();
+            println!("[A] nvme path: {path:?}");
+
+            let fio = FioBuilder::new()
+                .with_fio_binary(
```
Comment (dsharma-dc) [inline io-engine/tests/nexus_rebuild_partial_loop.rs]: Why do we have the need to provide fio binary explicitly, to rule out any fio bugs in default 3.33 ?

File: io-engine/tests/nexus_rebuild_partial_loop.rs (hunk 2 above)
Comment (dsavitskiy) [reply io-engine/tests/nexus_rebuild_partial_loop.rs:262]: On my machine, around one hour.

File: ci.nix (hunk 1 above)
Comment (dsavitskiy) [reply ci.nix]: Just in case it cannot be found, in order to prevent the version message from failing.

File: io-engine-tests/src/fio.rs (hunk 3 above)
Comment (dsavitskiy) [reply io-engine-tests/src/fio.rs]: Yes.

File: io-engine-tests/src/nexus.rs (hunk 4 above)
Comment (dsavitskiy) [reply io-engine-tests/src/nexus.rs]: Sorry, I forgot to clean up tmp debug stuff.

File: io-engine/tests/nexus_rebuild_partial_loop.rs (hunk 5 above)
Comment (dsavitskiy) [reply io-engine/tests/nexus_rebuild_partial_loop.rs]: Yes, Fio builder API for providing fio binary explicitly is needed for develpoment / debug purposes.
But this one is, again, I forgot to clean up.

File: ci.nix (hunk 1 above)
Comment (tiagolobocastro) [reply ci.nix]: I don't think this can happen. If fio fails to build then nix-shell will exit

File: io-engine/tests/nexus_rebuild_partial_loop.rs (hunk 2 above)
Comment (tiagolobocastro) [reply io-engine/tests/nexus_rebuild_partial_loop.rs:262]: Can we make this configurable somehow? If we create a curated list of tests 1h per each could be too much? If we run as a soak we may want to run for longer periods as well, ex: 1d or even 1w

Comment (dsavitskiy) [reply io-engine/tests/nexus_rebuild_partial_loop.rs:262]: Volume size, number of FIO loops and probably some FIO params can be made configurable for this test.
We'd need to discuss with Blaise how to do that better (via env var?).

Race conditions happen rarely, so a large number of loops is needed to be sure.

File: ci.nix (hunk 1 above)
Comment (dsavitskiy) [reply ci.nix]: You're right, I'll remove this.

**Review by dsharma-dc — APPROVED** [review]

**Review by tiagolobocastro — APPROVED** [review]

**Comment by tiagolobocastro** [comment]
bors merge
